### PR TITLE
Plfm 6218

### PR DESCRIPTION
**Pull request by brucehoff**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImpl.java [hunk 1]
```diff
@@ -136,6 +139,38 @@
...
+	ss.state='approved' and
+	ss.modified_on = 
+	select(max(ss2.modified_on)
+	from s2, ss2
+	where 
+	s2.ar_id=s.ar_id and s2.rp_id=s.rp_id
+	and s2.id=ss2.s_id and ss2.state=approved
+	group by s2.ar_id, s2.rp_id)
+	order by ss.modified_on asc;
+	*/
```
Comment (jay-hodgson) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImpl.java]: Thank you for the comment with the easy to read sql!  To comprehend, I only had to go to the real sql to figure out what the table aliases referenced.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java [hunk 2]
```diff
@@ -281,9 +314,74 @@ public void testListSubmissions() {
...
+		result.setInstitution(rp.getInstitution());
+		result.setIntendedDataUseStatement(rp.getIntendedDataUseStatement());
+		result.setProjectLead(rp.getProjectLead());
+		result.setModifiedOn(new Date(modifiedOn));
+		return result;
+		
+	}
+	
+	@Test
+	public void testListInfoForApprovedSubmissions() {
```
Comment (jay-hodgson) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java:333]: I need to take some notes to understand this test, let's spend extra time on this section during our meeting this afternoon.

Comment (brucehoff) [reply lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java:333]: Yes, this is the crucial part of the code review.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImpl.java (hunk 1 above)
Comment (brucehoff) [reply lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImpl.java]: I just replaced the hand written pseudo-code with the actual generated SQL.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmission.java [hunk 3]
```diff
@@ -212,9 +233,30 @@ public MigrationType getMigratableTableType() {
...
 	public MigratableTableTranslation<DBOSubmission, DBOSubmission> getTranslator() {
-		return new BasicMigratableTableTranslation<DBOSubmission>();
+		return new MigratableTableTranslation<DBOSubmission,DBOSubmission>(){
+
+			@Override
+			public DBOSubmission createDatabaseObjectFromBackup(DBOSubmission backup) {
+				if (backup.getResearchProjectId()==null) {
+					try {
+						Submission submission = (Submission)JDOSecondaryPropertyUtils.decompressObject(X_STREAM, 
+								submissionSerialized);
```
Comment (jay-hodgson) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmission.java]: backup.getSubmissionSerialized()

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionTest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		backup.setCreatedOn(3L);
+		backup.setDataAccessRequestId(4L);
+		backup.setEtag("etag");
+		backup.setId(5L);
+		backup.setSubmissionSerialized(JDOSecondaryPropertyUtils.compressObject(X_STREAM, submission));
+		
+		// Note, we have NOT set the research project id in backup
+		assertNull(backup.getResearchProjectId());
+		
+		DBOSubmission dbo = backup.getTranslator().createDatabaseObjectFromBackup(backup);
```
Comment (jay-hodgson) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionTest.java]: Replace backup with new DBOSubmission() to catch bug.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java [hunk 5]
```diff
@@ -119,16 +140,24 @@ public void before() {
...
 		}
 		if (researchProject != null) {
 			researchProjectDao.delete(researchProject.getId());
 		}
+		if (researchProject2 != null) {
+			researchProjectDao.delete(researchProject2.getId());
+		}
 		if (accessRequirement != null) {
 			accessRequirementDAO.delete(accessRequirement.getId().toString());
 		}
```
Comment (jay-hodgson) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java:163]: clean up accessRequirement2

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java [hunk 6]
```diff
@@ -281,9 +314,74 @@ public void testListSubmissions() {
...
+		List<SubmissionInfo> expected = ImmutableList.of(dto3Info, dto4Info);
+
+		// method under test
+		List<SubmissionInfo> actual = submissionDao.listInfoForApprovedSubmissions(accessRequirement.getId().toString(), 10, 0);
+		
+		assertEquals(expected, actual);
+		
+		
+		// check that pagination works right:  If I get the second page of size *one*, I should just get dto4
+		actual = submissionDao.listInfoForApprovedSubmissions(accessRequirement.getId().toString(), 1, 1);
```
Comment (jay-hodgson) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOSubmissionDAOImplTest.java:389]: Could you add a test to get a limit of 1 and offset of 0, and make sure you get dto3Info?

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**